x-pack/plugins/endpoint/server/routes/resolver/queries/base.ts

```diff
@@ -7,50 +7,91 @@
+      return { query: this.legacyQuery(this.endpointID, idsArray), index: legacyEventIndexPattern };
     }
-    return this.query(ids, this.indexPattern);
+    return { query: this.query(idsArray), index: this.indexPattern };
+  }
+
+  private buildSearch(ids: string | string[]) {
+    const { query, index } = this.buildQuery(ids);
+    return {
+      body: query,
+      index,
+    };
+  }
+
+  protected static getResults(response: SearchResponse<ResolverEvent>): ResolverEvent[] {
+    return response.hits.hits.map(hit => hit._source);
   }
 
-  async search(client: IScopedClusterClient, ...ids: string[]) {
-    return this.postSearch(await client.callAsCurrentUser('search', this.build(...ids)));
+  /**
+   * Builds a multi search representation for this query
+   *
+   * @param ids a single more multiple (e.g. entity_id or unique_pid) to search for in the query
+   * @returns an array of header and body pairs that represents a multi search
+   * https://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html
+   * https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/msearch_examples.html
+   */
+  buildMSearch(ids: string | string[]): JsonObject[] {
+    const { query, index } = this.buildQuery(ids);
+    return [{ index }, query];
   }
 
-  protected postSearch(response: SearchResponse<ResolverEvent>): PaginatedResults {
-    return paginatedResults(response);
+  /**
+   * Searches ES for the specified ids.
+   *
+   * @param client a client for searching ES
+   * @param ids a single more multiple unique node ids (e.g. entity_id or unique_pid)
+   */
+  async search(client: IScopedClusterClient, ids: string | string[]): Promise<T> {
+    const res: SearchResponse<ResolverEvent> = await client.callAsCurrentUser(
+      'search',
+      this.buildSearch(ids)
+    );
+    return this.formatResponse(res);
   }
 
-  protected abstract legacyQuery(
-    endpointID: string,
-    uniquePIDs: string[],
-    index: string
-  ): JsonObject;
-  protected abstract query(entityIDs: string[], index: string): JsonObject;
+  /**
+   * Builds a query to search the legacy data format.
+   *
+   * @param endpointID a unique identifier for a sensor
+   * @param uniquePIDs array of unique process IDs to search for
+   * @returns a query to use in ES
+   */
+  protected abstract legacyQuery(endpointID: string, uniquePIDs: string[]): JsonObject;
+
+  /**
+   * Builds a query to search for events in ES.
+   *
+   * @param entityIDs array of unique identifiers for events treated as nodes
+   */
+  protected abstract query(entityIDs: string[]): JsonObject;
+
+  /**
+   * Translates the response from executing the derived class's query into a structured object
+   *
+   * @param response a SearchResponse from ES resulting from executing this query
+   * @returns the translated ES response into a structured object
+   */
+  public abstract formatResponse(response: SearchResponse<ResolverEvent>): T;
```

Comment:
The reason this is `public` instead of `protected` is because it will need to be called by code that is performing a `msearch` which is not implemented in this class. The actual `msearch` implementation can take multiple queries so the caller will handle figuring out which responses map to which queries to do the `formatResponse` call.